common.mk

```diff
@@ -114,69 +114,68 @@ generator_temp: $(SCALA_SOURCES) $(sim_files) $(SCALA_BUILDTOOL_DEPS) $(EXTRA_GE
 		--top-module $(MODEL_PACKAGE).$(MODEL) \
 		--legacy-configs $(CONFIG_PACKAGE):$(CONFIG) \
 		$(EXTRA_CHISEL_OPTIONS))
+	sed -i '1d' $(ANNO_FILE)
+	echo "[" >> t.json
+	echo "{" >> t.json
+	echo "  \"class\": \"sifive.enterprise.firrtl.ModuleHierarchyAnnotation\"," >> t.json
+	echo "  \"filename\": \"./mod-he.json\"" >> t.json
+	echo "}," >> t.json
+	mv $(ANNO_FILE) t2.json
+	cat t.json t2.json > $(ANNO_FILE)
+	rm t.json t2.json
 
 .PHONY: firrtl
 firrtl: $(FIRRTL_FILE)
 
 #########################################################################################
 # create verilog files rules and variables
 #########################################################################################
-REPL_SEQ_MEM = --infer-rw --repl-seq-mem -c:$(MODEL):-o:$(TOP_SMEMS_CONF)
-HARNESS_CONF_FLAGS = -thconf $(HARNESS_SMEMS_CONF)
-
-TOP_TARGETS = $(TOP_FILE) $(TOP_SMEMS_CONF) $(TOP_ANNO) $(TOP_FIR) $(sim_top_blackboxes)
-HARNESS_TARGETS = $(HARNESS_FILE) $(HARNESS_SMEMS_CONF) $(HARNESS_ANNO) $(HARNESS_FIR) $(sim_harness_blackboxes)
+CIRCT_TARGETS = $(VSRC_SMEMS_CONF) $(VSRC_MODH_JSON)
 
 # DOC include start: FirrtlCompiler
 # NOTE: These *_temp intermediate targets will get removed in favor of make 4.3 grouped targets (&: operator)
 .INTERMEDIATE: firrtl_temp
-$(TOP_TARGETS) $(HARNESS_TARGETS): firrtl_temp
+$(CIRCT_TARGETS): firrtl_temp
 	@echo "" > /dev/null
 
 firrtl_temp: $(FIRRTL_FILE) $(ANNO_FILE) $(VLOG_SOURCES)
-	$(call run_scala_main,tapeout,barstools.tapeout.transforms.GenerateTopAndHarness,\
-		--allow-unrecognized-annotations \
-		--output-file $(TOP_FILE) \
-		--harness-o $(HARNESS_FILE) \
-		--input-file $(FIRRTL_FILE) \
-		--syn-top $(TOP) \
-		--harness-top $(VLOG_MODEL) \
-		--annotation-file $(ANNO_FILE) \
-		--top-anno-out $(TOP_ANNO) \
-		--top-dotf-out $(sim_top_blackboxes) \
-		--top-fir $(TOP_FIR) \
-		--harness-anno-out $(HARNESS_ANNO) \
-		--harness-dotf-out $(sim_harness_blackboxes) \
-		--harness-fir $(HARNESS_FIR) \
-		$(REPL_SEQ_MEM) \
-		$(HARNESS_CONF_FLAGS) \
-		--target-dir $(build_dir) \
-		--log-level $(FIRRTL_LOGLEVEL) \
-		$(EXTRA_FIRRTL_OPTIONS))
-	touch $(sim_top_blackboxes) $(sim_harness_blackboxes)
+	$(SCRATCH_HOME)/circt/build/bin/firtool \
```

Comment:
How does this work with Chisel aspects? Can you still inject FIRRTL passes through annotations?

Reply:
If my understanding above is correct, it seems like if you have your own custom FIRRTL passes, you'd need to run them separately, and if they work, make a Phase that extends CIRCT and compile that into your own custom firtool binary?

Reply:
Ah, some documentation: https://github.com/llvm/circt/blob/eba1bb5b81578e83a6457bf2cf32b94dda00b813/docs/Dialects/FIRRTL/RationaleFIRRTL.md and https://github.com/llvm/circt/blob/16f18f40b003023b8abb452ee85e4c5255107201/docs/Dialects/FIRRTL/FIRRTLAnnotations.md
From the first link:
> The FIRRTL dialect in CIRCT is designed to provide a drop-in replacement for the SFC for the subset of FIRRTL IR that is produced by Chisel and in common use.
It explicitly lacks `Fixed` and `Interval` datatype support (ouch, for DSP ppl - this might be a dealbreaker, such as for @milovanovic?), and from the second link seems to statically support the list of annotations in the second link. So yeah, looks like custom passes/annotations need to be run separately.

Reply:
That's roughly my understanding. You need to re-write any Scala FIRRTL passes into C++ and compile them into your binary. I do wonder if there is any way to inject passes (which was an old feature that the Scala FIRRTL compiler supported).

Reply:
Yeah, in addition to figuring out if passes can be injected, it would also be informative to know whether/if there are any migration rules for passes from SFC to CIRCT.

Reply:
Also, if CIRCT's subset is too small, can we keep targeting SFC as a fallback instead of replacing it entirely?

Reply:
I was also curious about the topic of adding custom passes to CIRCT, and asked about it here: https://discourse.llvm.org/t/writing-custom-firrtl-passes/62216

The answer basically boiled down to: it's not currently supported, but that's only because of lack of work on the feature. Ideally custom passes could be upstreamed into CIRCT, but for those that can't, some work needs to be put in to develop a system for user-defined passes. Right now, the only way to make those happen is by forking CIRCT and compiling the `firtool` binary yourself after writing your custom pass.

Reply:
I think my working goal right now is to have the following flow: Chisel > SFC (output .for) > CIRCT (output .sv). Here the SFC would do a couple of different things depending on if there are custom passes/unsupported constructs in the .fir. It would basically be a preprocessing step which would run extra passes (if found)(this would potentially lower  CHIRRTL and take a long time), and/or resolve unsupported FIRRTL constructs (again probably lowering the .fir)

Reply:
Hold on--is the main time-consuming step in SFC is lowering to any form of FIRRTL, or the actual transformation actions while the circuit is in FIRRTL, or the lowering of FIRRTL to Verilog? Seems like definitely not the last one, so wouldn't your proposed flow barely save time?

Separately, if unsupported constructs are limited to the mentioned data types, you should be able to do this pre-processing in Chisel?

Reply:
My assumption is that in most cases the SFC will just be a no-op, passing the FIRRTL directly to CIRCT. However, if there are special types (i.e. `Interval`, etc) or if there are custom passes that should be run then SFC will do all of the lowering (time-consuming) (and CIRCT will just handle the low FIRRTL -> Verilog). This flow hopefully will have everyone use CIRCT + SFC instead of a make flag that enables SFC v.s. CIRCT.

I'm not sure how you do Chisel pre-processing for converting a `Interval`/`Fixed` type to a "normal" `UInt` (or other supported type). Can you clarify what you were thinking here?

Reply:
> This flow hopefully will have everyone use CIRCT + SFC instead of a make flag that enables SFC v.s. CIRCT.

Got it, sounds good.

Sorry, I meant just detection of the unsupported constructs, not conversion to a supported datatype. But your proposal of running a lightweight SFC pass that determines whether it passes high FIRRTL to CIRCT or not would do the job.

Reply:
Hey. Feel free to ping me directly if you all have questions about this. I don't always see these discussions.

> How does this work with Chisel aspects?

Aspects that run at Chisel-time and generate annotations still work as those are entirely in Chisel. There's no plan to support `InjectingAspects` (annotations that append blobs of FIRRTL into a module) as that is viewed as a pretty janky API.

Re: annotations. That `.md` may be slightly out of date. The guaranteed up-to-date list of accepted annotations in in this header file: https://github.com/llvm/circt/blob/main/include/circt/Dialect/FIRRTL/AnnotationDetails.h Annotations are currently allow-listed. `firtool` will error on an annotation it doesn't understand (unless you run with `--disable-annotation-unknown` as this PR is doing).

> It explicitly lacks Fixed and Interval datatype support...

We're on the fence about `Fixed`. It's not that much work to add it, we just don't use it. I'm not totally against an open source contribution to add it to CIRCT, but start a discussion before you go this route. Jack has also thought that `Fixed` may be entirely expressible in Chisel without relying on FIRRTL. This would be ideal. There are other discussions about user-defined types in FIRRTL which would be another way that this could show up.

`Intervals` is another story as that was always very experimental, but did get usage at Berkeley. Value range analysis is bread and butter for compilers and there is already some work in MLIR for doing this with the `arith` dialect (which is used for ML/software). I expect this will show up in CIRCT eventually, just not in FIRRTL.

> Hold on--is the main time-consuming step in SFC is lowering to any form of FIRRTL, or the actual transformation actions while the circuit is in FIRRTL, or the lowering of FIRRTL to Verilog? Seems like definitely not the last one, so wouldn't your proposed flow barely save time?

That's the main drawback of mixing a Scala-based FIRRTL Compiler (SFC) flow with an MLIR-based FIRRTL Compiler (MFC/CIRCT) flow for the purposes of running custom transforms. In my experience, almost all non-trivial custom transforms are written against Low FIRRTL because it's extremely difficult to work with higher level FIRRTL in the SFC. You then wind up running almost the whole compilation through SFC for minimal gain (other than it being an incremental step towards moving to MFC).

One new development that is happening is MFC will support Chisel's [Trace API](https://www.chisel-lang.org/api/latest/chisel3/experimental/Trace$.html). This will enable you to get the final path of anything you annotate. This is expected to be sufficient enough to replace many annotations/custom passes that are used to emit information from a design, e.g., TCL generation. There were also a large number of SiFive custom passes that made there way into CIRCT (without accompanying Chisel APIs). There may be overlap with existing custom SFC passes. (I see you've turned on `sifive.enterprise.firrtl.ModuleHierarchyAnnotation`. 😉 )

Reply:
Thanks for the clarification! I agree that using SFC for custom transforms will most likely remove all the benefits of the CIRCT compiler. However, I expect that most people don't need custom transforms so this will still be a benefit.